File: android/app/src/main/java/com/prabotics/rapidrop/security/Curve25519.kt
package com.prabotics.rapidrop.security

import java.math.BigInteger
import java.security.SecureRandom



object Curve25519 {
    private val P: BigInteger = BigInteger.valueOf(2).pow(255).subtract(BigInteger.valueOf(19))
    private val A24: BigInteger = BigInteger.valueOf(121665)
    private val BASE_POINT: ByteArray = ByteArray(32).apply { this[0] = 9 }

    fun generatePrivateKey(random: SecureRandom = SecureRandom()): ByteArray {
        val key = ByteArray(32)
        random.nextBytes(key)
        return key
    }

    fun computePublicKey(privateKey: ByteArray): ByteArray {
        return scalarMult(privateKey, BASE_POINT)
    }

    fun computeSharedSecret(privateKey: ByteArray, remotePublicKey: ByteArray): ByteArray {
        return scalarMult(privateKey, remotePublicKey)
    }

    private fun clamp(k: ByteArray): ByteArray {
        val clamped = k.copyOf(32)
        clamped[0] = (clamped[0].toInt() and 248).toByte()
        clamped[31] = (clamped[31].toInt() and 127).toByte()
        clamped[31] = (clamped[31].toInt() or 64).toByte()
        return clamped
    }

    private fun scalarMult(scalar: ByteArray, uBytes: ByteArray): ByteArray {
        val clamped = clamp(scalar)
        val kInt = decodeLittleEndian(clamped)
        val u = decodeLittleEndian(uBytes).mod(P)

        val x1 = u
        var x2 = BigInteger.ONE
        var z2 = BigInteger.ZERO
        var x3 = u
        var z3 = BigInteger.ONE
        var swap = 0

        for (t in 254 downTo 0) {
            val kt = kInt.testBit(t)
            val ktInt = if (kt) 1 else 0
            swap = swap xor ktInt

            if (swap == 1) {
                var dummy = x2
                x2 = x3
                x3 = dummy

                dummy = z2
                z2 = z3
                z3 = dummy
            }
            swap = ktInt

            val a = x2.add(z2).mod(P)
            val aa = a.multiply(a).mod(P)
            val b = x2.subtract(z2).mod(P)
            val bb = b.multiply(b).mod(P)
            val e = aa.subtract(bb).mod(P)
            val c = x3.add(z3).mod(P)
            val d = x3.subtract(z3).mod(P)
            val da = d.multiply(a).mod(P)
            val cb = c.multiply(b).mod(P)

            val daPlusCb = da.add(cb).mod(P)
            x3 = daPlusCb.multiply(daPlusCb).mod(P)

            val daMinusCb = da.subtract(cb).mod(P)
            z3 = x1.multiply(daMinusCb.multiply(daMinusCb).mod(P)).mod(P)

            x2 = aa.multiply(bb).mod(P)
            val a24TimesE = A24.multiply(e).mod(P)
            z2 = e.multiply(aa.add(a24TimesE).mod(P)).mod(P)
        }

        if (swap == 1) {
            var dummy = x2
            x2 = x3
            x3 = dummy

            dummy = z2
            z2 = z3
            z3 = dummy
        }

        val zInv = z2.modInverse(P)
        val result = x2.multiply(zInv).mod(P)
        return encodeLittleEndian(result)
    }

    private fun decodeLittleEndian(bytes: ByteArray): BigInteger {
        val reversed = ByteArray(bytes.size)
        for (i in bytes.indices) {
            reversed[i] = bytes[bytes.size - 1 - i]
        }
        return BigInteger(1, reversed)
    }

    private fun encodeLittleEndian(value: BigInteger): ByteArray {
        val raw = value.toByteArray()
        val result = ByteArray(32)
        var srcPos = raw.size - 1
        var dstPos = 0
        while (srcPos >= 0 && dstPos < 32) {
            result[dstPos++] = raw[srcPos--]
        }
        return result
    }
}

File: windows/src/RapiDrop.Core/Security/Curve25519.cs
using System.Numerics;
using System.Security.Cryptography;

namespace RapiDrop.Core.Security;

public static class Curve25519
{
    private static readonly BigInteger P = BigInteger.Pow(2, 255) - 19;
    private static readonly BigInteger A24 = 121665;
    private static readonly byte[] BasePoint = new byte[32] { 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };

    public static byte[] GeneratePrivateKey()
    {
        byte[] key = new byte[32];
        RandomNumberGenerator.Fill(key);
        return key;
    }

    public static byte[] ComputePublicKey(byte[] privateKey)
    {
        return ScalarMult(privateKey, BasePoint);
    }

    public static byte[] ComputeSharedSecret(byte[] privateKey, byte[] remotePublicKey)
    {
        return ScalarMult(privateKey, remotePublicKey);
    }

    private static byte[] Clamp(byte[] k)
    {
        byte[] clamped = new byte[32];
        Buffer.BlockCopy(k, 0, clamped, 0, 32);
        clamped[0] &= 248;
        clamped[31] &= 127;
        clamped[31] |= 64;
        return clamped;
    }

    private static byte[] ScalarMult(byte[] scalar, byte[] uBytes)
    {
        byte[] clamped = Clamp(scalar);
        BigInteger kInt = DecodeLittleEndian(clamped);
        BigInteger u = Mod(DecodeLittleEndian(uBytes), P);

        BigInteger x1 = u;
        BigInteger x2 = BigInteger.One;
        BigInteger z2 = BigInteger.Zero;
        BigInteger x3 = u;
        BigInteger z3 = BigInteger.One;
        int swap = 0;

        for (int t = 254; t >= 0; t--)
        {
            int kt = (int)((kInt >> t) & 1);
            swap ^= kt;

            if (swap == 1)
            {
                (x2, x3) = (x3, x2);
                (z2, z3) = (z3, z2);
            }
            swap = kt;

            BigInteger a = Mod(x2 + z2, P);
            BigInteger aa = Mod(a * a, P);
            BigInteger b = Mod(x2 - z2, P);
            BigInteger bb = Mod(b * b, P);
            BigInteger e = Mod(aa - bb, P);
            BigInteger c = Mod(x3 + z3, P);
            BigInteger d = Mod(x3 - z3, P);
            BigInteger da = Mod(d * a, P);
            BigInteger cb = Mod(c * b, P);

            BigInteger daPlusCb = Mod(da + cb, P);
            x3 = Mod(daPlusCb * daPlusCb, P);

            BigInteger daMinusCb = Mod(da - cb, P);
            z3 = Mod(x1 * Mod(daMinusCb * daMinusCb, P), P);

            x2 = Mod(aa * bb, P);
            BigInteger a24TimesE = Mod(A24 * e, P);
            z2 = Mod(e * Mod(aa + a24TimesE, P), P);
        }

        if (swap == 1)
        {
            (x2, x3) = (x3, x2);
            (z2, z3) = (z3, z2);
        }

        BigInteger zInv = BigInteger.ModPow(z2, P - 2, P);
        BigInteger result = Mod(x2 * zInv, P);
        return EncodeLittleEndian(result);
    }

    private static BigInteger Mod(BigInteger x, BigInteger m)
    {
        BigInteger r = x % m;
        return r < 0 ? r + m : r;
    }

    private static BigInteger DecodeLittleEndian(byte[] bytes)
    {
        byte[] buffer = new byte[bytes.Length + 1];
        Buffer.BlockCopy(bytes, 0, buffer, 0, bytes.Length);
        buffer[bytes.Length] = 0;
        return new BigInteger(buffer);
    }

    private static byte[] EncodeLittleEndian(BigInteger value)
    {
        byte[] raw = value.ToByteArray();
        byte[] result = new byte[32];
        int count = Math.Min(raw.Length, 32);
        Buffer.BlockCopy(raw, 0, result, 0, count);
        return result;
    }
}
